G-API: broken compilation of samples (winpack, MSVS 2017)

[Nightly MSVS2017 build is broken](http://pullrequest.opencv.org/buildbot/builders/master-win64-vc15/builds/419).

[Winpack configuration](http://pullrequest.opencv.org/buildbot/builders/master_winpack-test-win64-vc15/builds/872) (samples against prebuilt binaries)

- MSVS 2017 has problem: 19.16.27035.0 (15.9.20), 19.16.27042.0 (15.9.25 - **latest**)
- MSVS 2015 is clean
- MSVS 2019 is clean (manual test). It has similar [bug report against compiler](https://developercommunity.visualstudio.com/content/problem/558844/error-c3546-there-are-no-parameter-packs-available.html).

relates #17502
/cc @dmatveev 

Error message:
```
C:\build\master_winpack-test-win64-vc15\build\distrib\opencv\build\include\opencv2/gapi/infer.hpp(111): error C3546: '...': there are no parameter packs available to expand [C:\build\master_winpack-test-win64-vc15\build\samples_build\cpp\example_tutorial_age_gender_emotion_recognition.vcxproj]
  C:\build\master_winpack-test-win64-vc15\build\distrib\opencv\sources\samples\cpp\tutorial_code\gapi\age_gender_emotion_recognition\age_gender_emotion_recognition.cpp(79): note: see reference to class template instantiation 'cv::GNetworkType<custom::Faces,std::function<cv::GMat (cv::GMat)>>' being compiled
C:\build\master_winpack-test-win64-vc15\build\distrib\opencv\build\include\opencv2/gapi/infer.hpp(87): error C3546: '...': there are no parameter packs available to expand [C:\build\master_winpack-test-win64-vc15\build\samples_build\cpp\example_tutorial_age_gender_emotion_recognition.vcxproj]
  C:\build\master_winpack-test-win64-vc15\build\distrib\opencv\sources\samples\cpp\tutorial_code\gapi\age_gender_emotion_recognition\age_gender_emotion_recognition.cpp(85): note: see reference to class template instantiation 'cv::GNetworkType<custom::AgeGender,std::function<custom::AGInfo (cv::GMat)>>' being compiled
```

--- a/modules/gapi/include/opencv2/gapi/infer.hpp
+++ b/modules/gapi/include/opencv2/gapi/infer.hpp
@@ -77,15 +77,6 @@ public:
 
     using ResultL = std::tuple< cv::GArray<R>... >;
     using APIList = std::function<ResultL(cv::GArray<cv::Rect>, Args...)>;
-
-    // APIList2 is also template to allow different calling options
-    // (GArray<cv::Rect> vs GArray<cv::GMat> per input)
-    template<class... Ts>
-    using APIList2 = typename std::enable_if
-        < cv::detail::valid_infer2_types< std::tuple<Args...>
-                                        , std::tuple<Ts...> >::value,
-          std::function<ResultL(cv::GMat, cv::GArray<Ts>...)>
-        >::type;
 };
 
 // Single-return-value network definition (specialized base class)
@@ -101,14 +92,16 @@ public:
 
     using ResultL = cv::GArray<R>;
     using APIList = std::function<ResultL(cv::GArray<cv::Rect>, Args...)>;
+};
 
-    // APIList2 is also template to allow different calling options
-    // (GArray<cv::Rect> vs GArray<cv::GMat> per input)
-    template<class... Ts>
-    using APIList2 = typename std::enable_if
-        < cv::detail::valid_infer2_types< std::tuple<Args...>
+// APIList2 is also template to allow different calling options
+// (GArray<cv::Rect> vs GArray<cv::GMat> per input)
+template<class Net, class... Ts>
+struct InferAPIList2 {
+    using type = typename std::enable_if
+        < cv::detail::valid_infer2_types< typename Net::InArgs
                                         , std::tuple<Ts...> >::value,
-          std::function<ResultL(cv::GMat, cv::GArray<Ts>...)>
+          std::function<typename Net::ResultL(cv::GMat, cv::GArray<Ts>...)>
         >::type;
 };
 
@@ -185,7 +178,7 @@ template<typename Net, typename... Args>
 struct GInferList2 final
     : public GInferList2Base
     , public detail::KernelTypeMedium< GInferList2<Net, Args...>
-                                     , typename Net::template APIList2<Args...> > {
+                                     , typename InferAPIList2<Net, Args...>::type > {
     using GInferList2Base::getOutMeta; // FIXME: name lookup conflict workaround?
 
     static constexpr const char* tag() { return Net::tag(); }

